### PR TITLE
RL math evaluation

### PR DESCRIPTION
**Pull request by AlexPiche**
## Fixes

- [x] Fix `count_tokens`. It was off by 4 tokens since it did not include the generation prompt.
- [x] Fix reward normalization when the reward is `reward - coef * kl`

## Math Evaluation

- [x] Use the same prompt as deepseek.
- [x] Copy paste the deepseek code.
- [x] Add MIT license for deepseek code.  

This PR aims to accommodate both GSM8k and MATH dataset and to reproduce the deepseek math scores for these 2 datasets. The PR also simplify the orchestrator script and fix a typo in finetune/rl.

| Model                | Dataset | TapeAgents | DeepSeek Math Paper |
|----------------------|---------|------------|----------------------|
| DeepSeekMath-RL 7B   | GSM8k   | 88.0%      | 88.2%               |
| DeepSeekMath-RL 7B   | MATH    | 50.6%      | 51.7%               |
| DeepSeekMath-Instruct 7B   | GSM8k   | 82.5%      | 82.9%               |
| DeepSeekMath-Instruct 7B   | MATH   |     46.1%  | 46.8%               |

### Difference in generations

vLLM sometimes produce different generations than the ones reported by deepseek at temperature 0. For example, for the same prompt it is sometimes the exact same generation, sometimes they differ e.g.

#### TapeAgents
```json
{
    "prompt": "User: For what real values of $x$ is $-4<x^{4}+4x^{2}<21$ satisfied? Express your answer in interval notation.\nPlease reason step by step, and put your final answer within \\boxed{}.\n\nAssistant:",
    "model_output": " First, we can rewrite the inequality as two separate inequalities:\n\\[x^{4}+4x^{2}<21 \\quad \\text{and} \\quad x^{4}+4x^{2}>-4.\\]\nLet's solve the first inequality $x^{4}+4x^{2}<21$.\nLet $y=x^{2}$, then the inequality becomes $y^{2}+4y<21$.\nRearranging terms, we get $y^{2}+4y-21<0$.\nThis is a quadratic inequality, and we can solve it by finding the roots of the quadratic equation $y^{2}+4y-21=0$.\nUsing the quadratic formula, we find that the roots are $y = \\frac{-4 \\pm \\sqrt{4^2 - 4(-21)}}{2} = \\frac{-4 \\pm \\sqrt{16+84}}{2} = \\frac{-4 \\pm \\sqrt{100}}{2} = \\frac{-4 \\pm 10}{2}$.\nSo the roots are $y = -7$ and $y = 3$.\nSince the coefficient of $y^2$ is positive, the graph of $y^2 + 4y - 21 = 0$ is a parabola opening upwards.\nTherefore, the inequality $y^2 + 4y - 21 < 0$ is satisfied when $-7 < y < 3$.\nSubstituting $y = x^2$ back in, we have $-7 < x^2 < 3$.\nTaking the square root of all parts of the inequality, we get $-\\sqrt{3} < x < -\\sqrt{7}$ and $\\sqrt{7} < x < \\sqrt{3}$.\nCombining these two intervals, we have the solution for the first inequality: $x \\in \\left(-\\sqrt{3}, -\\sqrt{7}\\right) \\cup \\left(\\sqrt{7}, \\sqrt{3}\\right)$.\n\nNow let's solve the second inequality $x^{4}+4x^{2}>-4$.\nLet $y=x^{2}$, then the inequality becomes $y^{2}+4y>-4$.\nRearranging terms, we get $y^{2}+4y+4>0$.\nThis is a quadratic inequality, and we can solve it by factoring the left side: $(y+2)^2 > 0$.\nThis inequality is always true for $y \\neq -2$.\nSince $y = x^2$, $x^2 \\neq -2$ for all real $x$.\nTherefore, the second inequality is true for all real $x$.\n\nCombining the solutions of the two inequalities, we find that the solution to the original inequality is $x \\in \\left(-\\sqrt{3}, -\\sqrt{7}\\right) \\cup \\left(\\sqrt{7}, \\sqrt{3}\\right)$.\nThe answer is: $\\boxed{(-\\sqrt{3},-\\sqrt{7})\\cup(\\sqrt{7},\\sqrt{3})}$",
    "prediction": [
        "(-\\sqrt{3},-\\sqrt{7})\\cup(\\sqrt{7},\\sqrt{3})"
    ],
    "accuracy": 0
}
```

#### DeepSeek

```json
{
        "prompt": "User: For what real values of $x$ is $-4<x^{4}+4x^{2}<21$ satisfied? Express your answer in interval notation.\nPlease reason step by step, and put your final answer within \\boxed{}.\n\nAssistant:",
        "model_output": " First, we can rewrite the inequality as two separate inequalities:\n\\[x^{4}+4x^{2}<21 \\quad \\text{and} \\quad x^{4}+4x^{2}>-4.\\]\nLet's solve the first inequality $x^{4}+4x^{2}<21$.\nLet $y=x^{2}$, then the inequality becomes $y^{2}+4y<21$.\nRearranging terms, we get $y^{2}+4y-21<0$.\nThis is a quadratic inequality, and we can solve it by finding the roots of the quadratic equation $y^{2}+4y-21=0$.\nUsing the quadratic formula, we have $y=\\frac{-4\\pm\\sqrt{4^{2}-4(1)(-21)}}{2(1)}=\\frac{-4\\pm\\sqrt{100}}{2}=\\frac{-4\\pm10}{2}$.\nSo the roots are $y=-7$ and $y=3$.\nSince the coefficient of $y^{2}$ is positive, the graph of $y^{2}+4y-21$ is a parabola opening upwards.\nThe inequality $y^{2}+4y-21<0$ is satisfied when $y$ is between the roots $-7$ and $3$.\nSo we have $-7<x^{2}<3$.\nTaking the square root of all parts of the inequality, we get $-\\sqrt{3}<x<\\sqrt{3}$.\nNow let's solve the second inequality $x^{4}+4x^{2}>-4$.\nLet $y=x^{2}$, then the inequality becomes $y^{2}+4y>-4$.\nRearranging terms, we get $y^{2}+4y+4>0$.\nThis simplifies to $(y+2)^{2}>0$.\nThis inequality is satisfied for all $y\\neq -2$.\nSince $y=x^{2}$, this means $x^{2}\\neq -2$.\nHowever, $x^{2}$ is always nonnegative, so there are no restrictions on $x$ from this inequality.\nCombining the solutions from both inequalities, we find that the solution to the original inequality is $- \\sqrt{3} < x < \\sqrt{3}$.\nIn interval notation, this is $\\boxed{(-\\sqrt{3}, \\sqrt{3})}$. The answer is: $(-\\sqrt{3},\\sqrt{3})$",
        "prediction": [
            "(-\\sqrt{3},\\sqrt{3})"
        ],
        "accuracy": 1
    }
```

### CODE REVIEW

File: conf/finetune/rl_llama31_8b.yaml [hunk 1]
```diff
@@ -24,7 +24,7 @@ seq_length: 4096
 load_as_bf16: True
 max_train_steps: 100000
 save_checkpoint_steps: ???
-optim: adafactor # FIXME: adamw runs OOM with accelerate
+optim: adamw
```
Comment (rizar) [inline conf/finetune/rl_llama31_8b.yaml:27]: no OOM after all?

File: conf/rl_gsm8k.yaml [hunk 2]
```diff
@@ -2,24 +2,28 @@ defaults:
   - finetune: rl_llama31_8b
   - _self_
 
+dataset_name: gsm8k
 n_workers: 32
 get_log_probs_workers: 1
 max_loops: 1
 test_every_n_iterations: 5
-model_path: meta-llama/Meta-Llama-3.1-8B-Instruct
+model_path: /mnt/llmd/base_models/deepseek-math-7b-instruct
```
Comment (rizar) [inline conf/rl_gsm8k.yaml:10]: can you please comment why we are starting with deepseek now?

File: examples/rl_gsm8k/cot_math_agent.py [hunk 3]
```diff
@@ -19,59 +18,44 @@
...
-COT_GUIDANCE = "Think step by step. When you know the answer to the question, provide it in the following format: The answer is: <number>"
-
 
 class Task(Observation):
     kind: Literal["task"] = "task"
     task: str
 
     def llm_view(self, indent: int | None = 2) -> str:
-        return f"{self.task} {COT_GUIDANCE}"
+        # Same prompt as https://github.com/deepseek-ai/DeepSeek-Math/blob/b8b0f8ce093d80bf8e9a641e44142f06d092c305/evaluation/run_subset_parallel.py#L26
```
Comment (rizar) [inline examples/rl_gsm8k/cot_math_agent.py:27]: Thanks for noting the source of the prompt in the code!

File: examples/rl_gsm8k/utils.py [hunk 4]
```diff
@@ -295,6 +297,18 @@ def launch_training(config_dir: str, config_name: str, accelerate_cfg_path: str,
     if num_gpus == 0:
         raise ValueError("No GPUs available for finetuning")
 
+
+    if not use_accelerate:
```
Comment (rizar) [inline examples/rl_gsm8k/utils.py:301]: why do we wish to support training without `accelerate`?

File: tapeagents/llms.py [hunk 5]
```diff
@@ -91,12 +91,18 @@ def make_training_text(self, prompt: Prompt, output: LLMOutput) -> TrainingText:
         pass
 
     def log_output(self, prompt: Prompt, message: LLMOutput, cached: bool = False):
+        prompt_length_tokens = self.count_tokens(prompt.messages)
+        if message.content:
+            output_length_tokens = self.count_tokens(prompt.messages + [{"role": "assistant", "content": message.content}]) - prompt_length_tokens
```
Comment (rizar) [inline tapeagents/llms.py:96]: for this particular line I'd like @ollmer to take a look

**Review by rizar — COMMENTED** [review]
Great step forward! A few things caught my eye:

- going back to adamw
- the mode of launching training without `accelerate`
- the switch to deepseek as the initial model
- the tricky new formula for computing the number of output tokens (@ollmer, what do you think)?

File: tapeagents/llms.py (hunk 5 above)
Comment (ollmer) [reply tapeagents/llms.py:96]: Yep, looks correct, addresses nuances about tokens separating different chat messages.

**Review by rizar — APPROVED** [review]
lgtm!

**Pull request merged**